Docker images

1138-4EB opened the pull request:
Note to maintainers: this is a slightly complex change, so I marked some items with checkboxes. Please, feel free to check those which are already done and/or those which you agree with.

---

In this PR, using some *reference* docker images for SpinalHDL is suggested. This is desirable because:

- Installing all the dependencies to use SpinalHDL is not straightforward. Providing a ready-to-use and multi-platform solution can help new users try and use SpinalHDL.
- Having a reference context/environment can help tracking/reproducing issues/bugs, as it allows to tell apart problems related to the project from installation problems in the host.
  - See [1138-4EB/issue-runner](https://github.com/1138-4EB/issue-runner)
- ATM, all the dependencies are built and installed in the CI system (Travis) each time a test is run (i.e. three times when each commit is pushed). Execution time can be reduced if all the dependencies are preinstalled. 

---

The proposed setup is as follows:

- [x] A new repo named SpinalHDL/docker is created. This is a clone/copy of [1138-4EB/SpinalHDL-docker](https://github.com/1138-4EB/SpinalHDL-docker).
- [x] A maintainer of SpinalHDL creates a docker user. Right now, I am the owner and single member of [hub.docker.com/u/spinalhdl](https://hub.docker.com/u/spinalhdl), but I am willing to give access or transfer it to @Dolu1990, @wifasoi or any other active member.
  - [x] Protected environment variables `DOCKER_USER` and `DOCKER_PASS` are set up in `travis-ci.com/SpinalHDL/docker` and `travis-ci.com/SpinalHDL/SpinalHDL`.
- [ ] This PR is accepted.
- [ ] PR SpinalHDL/VexRiscv#52 is accepted.

---

The outcome will be the following workflow:

- Each time a commit is pushed to SpinalHDL/docker, docker image `spinalhdl/dev` is built in Travis. This image includes GHDL, Verilator, iverilog, cocotb, VUnit and scala.
  - If successful, the image is pushed to [hub.docker.com/r/spinalhdl/dev](https://hub.docker.com/r/spinalhdl/dev/tags).
  - See example of the execution: https://travis-ci.com/1138-4EB/SpinalHDL-docker/builds/97910417
  - Alternatively, or as a complement, a cron job can be set up in travis, so that the image is built periodically, even when no new commits are pushed.
- Each time a commit is pushed to SpinalHDL/SpinalHDL, image [hub.docker.com/r/spinalhdl/dev](https://hub.docker.com/r/spinalhdl/dev/tags) is pulled and SpinalHDL is compiled and tested in a container.
  - If successful, two docker images are built and pushed to hub.docker.com:
    - [hub.docker.com/r/spinalhdl/spinalhdl](https://hub.docker.com/r/spinalhdl/spinalhdl/tags) is based on `spinalhdl/dev`, and SpinalHDL is installed in `/opt/SpinalHDL`.
    - [hub.docker.com/r/spinalhdl/riscv](https://hub.docker.com/r/spinalhdl/riscv/tags) is based on `spinalhdl/spinalhdl` and the RISCV toolchain is installed in `/opt/riscv`.

So:

- Users willing to test changes to the SpinalHDL code base, can use `spinalhdl/dev`.
- Users willing to use SpinalHDL to generate and simulate hardware, can use `spinalhdl/spinalhdl`.
- Users willing to use SpinalHDL to generate and simulate hardware, and/or build RISCV software for VexRiscv Socs, can use `spinalhdl/riscv`.

---

Note that in the proposed Dockerfile for `spinalhdl/dev` the versions of the dependencies have been updated:

- [x] A fixed commit SHA is used for GHDL: https://github.com/1138-4EB/SpinalHDL-docker/blob/master/Dockerfile#L37
- [x] Version 4.0008 of Verilator is used: https://github.com/1138-4EB/SpinalHDL-docker/blob/master/Dockerfile#L43
- [x] Version 10.2 of iverilog is used: https://github.com/1138-4EB/SpinalHDL-docker/blob/master/Dockerfile#L53

---

The following is a list of changes which are either relevant or which I am not sure about. Please, review them carefully. This is why I marked the PR as work in progress (WIP).

- [x] CLI options `-jvm-opts travis/jvmopts.compile` and `-jvm-opts travis/jvmopts.test` are removed. I think these are specific to Travis' VMs: https://github.com/SpinalHDL/SpinalHDL/blob/dev/.travis.yml#L16-L17 Certainly, I don't know what do they do.
- [x] I think that this fix is specific to Travis' VMs: https://github.com/SpinalHDL/SpinalHDL/blob/dev/.travis.yml#L27-L32. It is removed. We can add it again if we have any related problem.
- [x] I don't think this option is required: https://github.com/SpinalHDL/SpinalHDL/blob/dev/.travis.yml#L5-L7. It is removed.
- [x] ATM, SpinalHDL is being tested with `openjdk9`, `openjdk10` and `oraclejdk11`. However, after this PR is merged, a single version will be used: openjdk8. I think this should be discussed. On the one hand, in https://github.com/SpinalHDL/VexRiscv#dependencies it is stated that "*# JAVA JDK 8. Do not try with JDK >= 9*", so I am not sure about which versions are supported in any or all of the repos in this project. On the other hand, I don't know how much does SpinalHDL depend on the underlying java version.
- [ ] The proposed scheme includes a single tag for each image (`dev`, `spinalhdl` and `riscv`). All of them are based on Ubuntu 16.04 Xenial and include Java JDK 8. Related to the previous item, multiple images can be created, all of them based on Ubuntu Xenial and including different versions of Java JDK. Alternatively, other images can be created with different base systems, e.g. Ubuntu 18.04 with Java JDK 9. I think that it'd be interesting to have a single reference image (probably based on Debian Stretch), but it is up to the maintainers to set a list of interesting reference platforms.
- [x] In this PR, caching is disabled in travis. Since compilation and tests are executed in docker containers, these are not shared with the host VM by default. However, if this feature is considered relevant, `.ivy2/cache` and `.sbt/boot/` can be bind to the host, so that caching can be supported.

---

In the current proposal, docker images are expected to be *rolling images**. I.e., no tags are used, just `latest`. However, it is possible to adapt this PR in order to tag images with a semver when a tagged commit (a release) is pushed to GitHub. This would allow users to pull either stable images or the one corresponding to the development branch.

---

References:

- https://github.com/plex1/SpinalDev
- https://github.com/umarcor/dbhi

1138-4EB commented:
@plex1, you are fast! I was just editing the comment to add a reference to your repo, which is the one I used as a base last summer. Certainly, that's why the proposed image is based on Ubuntu Xenial. Thank you!

I did not mention any GUI tool here, because I want to focus on compilation and tests first. But I think that having some *official* image would help other users (such as you and me) to build environments on top of them. I.e. [umarcor/dbhi](https://github.com/umarcor/dbhi) includes DBM tools and GtkWave, while [plex1/SpinalDev](https://github.com/plex1/SpinalDev) includes icestorm tools and IDEs.

Dolu1990 commented:
Hi ^^

I'm a newbie with docker, i cloned your dockerimage repo, builded it (sudo docker build --tag=friendlyhello .
)

But then when i run it (sudo docker run friendlyhello) nothing realy happend, i'm probably doing something wrong XD

Can you tell me how to process ? 
Thanks :)

1138-4EB commented:
Hi @Dolu1990! The image you built, `friendlyhello`, is equivalent to `spinalhdl/dev`. So, instead of

``` bash
sudo docker build --tag=friendlyhello .
sudo docker run friendlyhello
```

you could have done

``` bash
sudo docker pull spinalhdl/dev
sudo docker run spinalhdl/dev
```

or even 

``` bash
sudo docker run spinalhdl/dev
```

directly, because the image is pulled automatically if not available on the host.

This is just for you to know that you (as a user) don't need to build the image locally. Travis does it for you: https://travis-ci.com/1138-4EB/SpinalHDL-docker/builds/97910417

---

> But then when i run it (sudo docker run friendlyhello) nothing realy happend, i'm probably doing something wrong XD

First off, I would suggest that you have a look at https://docs.docker.com/install/linux/linux-postinstall/. Precisely, you should add your user to the `docker` group. Then, log out and log in again, or restart your host. This will allow you to execute docker commands without `sudo`.

Nevertheless, when you do `docker run <any_image>`, a container is created and the default application inside it is executed (which is typically bash). However, since you didn't provide any input/output, it just ends in a few seconds. You can see that the container was created and exited:

``` bash
# docker ps -a
CONTAINER ID        IMAGE               COMMAND                  CREATED             STATUS                      PORTS                  NAMES
d4e473498ddb        spinalhdl/dev       "/bin/bash"              5 seconds ago       Exited (0) 4 seconds ago                           tender_jackson
```

So, let's remove it and start again:

``` bash
# docker rm -f tender_jackson
```

NOTE: `tender_jackson` is the name that was given to the container I create. I should check the name in table output by `docker ps -a`.

---

We want to make something useful, so, we will add some runtime parameters to the container:

``` bash
# docker run --rm -it spinalhdl/dev
root@23a87007154d:/work#
```

Argument `--rm` will remove the container as soon as it exits. Arguments `-i` and `-t` will redirect our keyboard input to the container, and will forward the stdout/stderr of the container to the shell. As a result, you can see that the prompt changed, and we are now inside the container (a kind of temporal VM). You can open a different shell in the host and execute `docker ps -a` to see the STATUS.

Inside the container, you can check that all the dependencies are available. E.g.:

``` bash
root@23a87007154d:/work# ghdl --version
GHDL 0.36-dev (20181129-161-g3c30e3b) [Dunoon edition]
 Compiled with GNAT Version: 5.4.0
 llvm code generator
Written by Tristan Gingold.

Copyright (C) 2003 - 2015 Tristan Gingold.
GHDL is free software, covered by the GNU General Public License.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
root@23a87007154d:/work# verilator --version
Verilator 4.008 2018-12-01 rev UNKNOWN_REV
root@23a87007154d:/work#
```

Certainly, you can do whatever you want here. E.g., download SpinalHDL, build and test it:

``` bash
root@c24893adac28:/work# git clone https://github.com/SpinalHDL/SpinalHDL
Cloning into 'SpinalHDL'...
remote: Enumerating objects: 28, done.
remote: Counting objects: 100% (28/28), done.
remote: Compressing objects: 100% (23/23), done.
remote: Total 39484 (delta 2), reused 16 (delta 1), pack-reused 39456
Receiving objects: 100% (39484/39484), 13.24 MiB | 10.62 MiB/s, done.
Resolving deltas: 100% (17563/17563), done.
Checking connectivity... done.
root@c24893adac28:/work# cd SpinalHDL/
root@c24893adac28:/work/SpinalHDL# sbt clean
Getting org.scala-sbt sbt 1.2.7  (this may take some time)...
...
```

When you are done with tinkering, just exit. The container will be closed and removed:

``` bash
root@23a87007154d:/work# exit
exit
# docker ps -a
CONTAINER ID        IMAGE               COMMAND             CREATED             STATUS              PORTS               NAMES
```

---

The procedure above is interesting to have a first contact. But, all the content inside the container is temporal, so you need to push the changes you do to the code, because everything is removed when you exit. Let's do it the other way: clone the repo on the host, and bind the folder to the container:

``` bash
# git clone https://github.com/SpinalHDL/SpinalHDL
...
# cd SpinalHDL
# docker run --rm -it -v $(pwd):/src -w /src spinalhdl/dev
root@b94776140a26:/src# sbt clean
Getting org.scala-sbt sbt 1.2.8  (this may take some time)...
...
root@b94776140a26:/src# exit
```

Much better. Now, folder `SpinalHDL` in the host is bind to `/src` inside the container. Hence, we can either use GUI editors/IDEs in the host, or vim/nano inside the container. Also, when the container is removed, changes to the content in the bind folder will persist.

NOTE: I added `-w /src` to set the workdir inside the container to be the root of the folder I bind, instead of `~`. This is not required, but useful.

---

Yet, the procedure is interactive, i.e., you need to create the container and introduce commands one by one. We can make it easier by telling explicitly which commands to execute:

``` bash
# docker run --rm -it -v $(pwd):/src -w /src spinalhdl/dev "sbt clean"
Getting org.scala-sbt sbt 1.2.8  (this may take some time)...
...
```

Last, if you want to tell several commands explicitly, it is easier to write them in a script file:

``` bash
# Content of myscript.sh
sbt clean
sbt compile
sbt test
```

``` bash
# docker run --rm -it -v $(pwd):/src -w /src spinalhdl/dev "$(cat path/to/myscript.sh)"
Getting org.scala-sbt sbt 1.2.8  (this may take some time)...
...
```

If the script is located in the root of the repo, it is available inside the container, and we can call it directly:

``` bash
# docker run --rm -it -v $(pwd):/src -w /src spinalhdl/dev ./myscript.sh"
Getting org.scala-sbt sbt 1.2.8  (this may take some time)...
...
```

You will find that this is exactly what this PR executes each time a commit is pushed to Travis. See:

- https://github.com/1138-4EB/SpinalHDL/blob/dockerize/.travis.yml#L10
- https://github.com/1138-4EB/SpinalHDL/blob/dockerize/travis.sh#L166-L167
- https://github.com/1138-4EB/SpinalHDL/blob/dockerize/travis.sh#L163-L164
- https://github.com/1138-4EB/SpinalHDL/blob/dockerize/travis.sh#L96-L121
  - https://github.com/1138-4EB/SpinalHDL/blob/dockerize/travis.sh#L100
  - https://github.com/1138-4EB/SpinalHDL/blob/dockerize/travis.sh#L118

Summarizing, executing `docker run --rm -itv $(pwd):/src -w /src spinalhdl/dev ./travis.sh -t` will produce the same output as executing `./travis.sh -t` in the host. But, in the former, everything is done inside a `spinalhdl/dev` container, so no deps are required besides docker.

NOTE: I suppose that you have most (if not all) of the tools already installed on your host. You can give this a try at play-with-docker.com, where you can create empty nodes. 

---

Since you are new to docker, and because this PR includes some not docker specific content (ANSI colors, some helper headers for travis, management of docker hub credentials...), let's take it easy. Try to use it as a user/developer of SpinalHDL, not a maintainer. Ask about any other issues you find. When you feel comfortable with docker, we'll move forward.

Dolu1990 commented:
@1138-4EB Thank you very much for all the informations !
Got it working :)

Dolu1990 commented:
Sorry, i was underwater XD.

So, i checked some most of the boxes.

There is the following points to discuse : 

1)
> The master branch of GHDL is used: https://github.com/1138-4EB/SpinalHDL-docker/blob/master/Dockerfile#L36

Never used the master branch of another project for regressions. It often that they were broken for a little while, and all the regression chain fail. Also, it is realy hard sometime to figure out why the thing broke.

So always stick to a working tag :)

2)

> ATM, SpinalHDL is being tested with openjdk9, openjdk10 and oraclejdk11. However, after this PR is merged, a single version will be used: openjdk8. I think this should be discussed. On the one hand, in https://github.com/SpinalHDL/VexRiscv#dependencies it is stated that "# JAVA JDK 8. Do not try with JDK >= 9", so I am not sure about which versions are supported in any or all of the repos in this project. On the other hand, I don't know how much does SpinalHDL depend on the underlying java version.

Now all JDK >= 8 are ok (since i updated the SBT version). The comment on VexRiscv was outdated :) 

So, it's fine if the sbt clean test is run only on a single JDK, but at least, an sbt clean compile should run on all JDK >= 8. It is sometime quite easy to use old parts of the java API which were removed from recent JDK.

Would that be possible ? 
sudo update-alternatives --config java
sudo update-alternatives --config javac


> The proposed scheme includes a single tag for each image (dev, spinalhdl and riscv). All of them are based on Ubuntu 16.04 Xenial and include Java JDK 8. Related to the previous item, multiple images can be created, all of them based on Ubuntu Xenial and including different versions of Java JDK. Alternatively, other images can be created with different base systems, e.g. Ubuntu 18.04 with Java JDK 9. I think that it'd be interesting to have a single reference image (probably based on Debian Stretch), but it is up to the maintainers to set a list of interesting reference platforms.

Personnaly, i'm on Mint 18 (Xenial) , JDK8 :)

1138-4EB commented:
> Sorry, i was underwater XD.

This time, I was busy...
 
> Never used the master branch of another project for regressions. It often that they were broken for a little while, and all the regression chain fail. Also, it is realy hard sometime to figure out why the thing broke.
> 
> So always stick to a working tag :)

Although this is true in general, GHDL is quite special in this regard. It is a rapidly evolving project, as many VHDL2018 features are being added or fixed. But releases (tags) are only published once a year, or so. Also, there is no development branch different from master. Nonetheless, the master branch is quite stable.

Therefore, if you want to stick to a working tag, but avoid using an outdated version, I suggest to use the commit SHA of the latest passing build in [travis-ci.org/ghdl/ghdl](https://travis-ci.org/ghdl/ghdl/builds). So, ATM, it would be `git checkout -b spinal 50da90f509aa6de2961f1795af0be2452bc2c6d9`. Is this OK?

> Now all JDK >= 8 are ok (since i updated the SBT version). The comment on VexRiscv was outdated :)

Good to know.
 
> So, it's fine if the sbt clean test is run only on a single JDK, but at least, an sbt clean compile should run on all JDK >= 8.

Ok. This will need slightly rethinking the modifications to `.travis.yml`. I'll update this PR.

> It is sometime quite easy to use old parts of the java API which were removed from recent JDK.
> 
> Would that be possible ?
> sudo update-alternatives --config java
> sudo update-alternatives --config javac

I am neither a Java user nor a Ubuntu user, so I am lost here. Could you please elaborate to help me understand?

> Personnaly, i'm on Mint 18 (Xenial) , JDK8 :)

Then, are you ok with keeping Ubuntu Xenial as the single docker platform (for now)? Unfortunately, there is no official Mint docker image, so I think that Ubuntu Xenial is the closest, provided that the image does not include any desktop environment.

1138-4EB commented:
Regarding

> A maintainer of SpinalHDL creates a docker user. Right now, I am the owner and single member of [hub.docker.com/u/spinalhdl](https://hub.docker.com/u/spinalhdl), but I am willing to give access or transfer it to @Dolu1990, @wifasoi or any other active member.
>
>    * Protected environment variables `DOCKER_USER` and `DOCKER_PASS` are set up in `travis-ci.com/SpinalHDL/docker` and `travis-ci.com/SpinalHDL/SpinalHDL`.

@Dolu1990, did you create the docker user and/or fork my docker repo to SpinalHDL/docker?

Dolu1990 commented:
Hi :)

> Although this is true in general, GHDL is quite special in this regard. It is a rapidly evolving project, as many VHDL2018 features are being added or fixed. But releases (tags) are only published once a year, or so. Also, there is no development branch different from master. Nonetheless, the master branch is quite stable.
Therefore, if you want to stick to a working tag, but avoid using an outdated version, I suggest to use the commit SHA of the latest passing build in [travis-ci.org/ghdl/ghdl](https://travis-ci.org/ghdl/ghdl/builds). So, ATM, it would be `git checkout -b spinal 50da90f509aa6de2961f1795af0be2452bc2c6d9`. Is this OK?

Realy, i already losed days on broken master branches of other repo ^.^
Yes that's ok

> I am neither a Java user nor a Ubuntu user, so I am lost here. Could you please elaborate to help me understand?

Basicaly, when you have multiple java version installated, this command change the one used by default.

> Then, are you ok with keeping Ubuntu Xenial as the single docker platform (for now)? Unfortunately, there is no official Mint docker image, so I think that Ubuntu Xenial is the closest, provided that the image does not include any desktop environment.

That's good for me ^^, but just one thing, "does not include any desktop environment", does that mean that we can't run Intellij from the image ? Would it be possible to have IDE like Intellij / eclipse in the docker image ?


> @Dolu1990, did you create the docker user and/or fork my docker repo to SpinalHDL/docker?

I created an docker acount : dolu1990
You can add it :)

1138-4EB commented:
> Hi :)

Hi again!

> Yes that's ok

Now a commit SHA is used.

> Basicaly, when you have multiple java version installated, this command change the one used by default.

I understand it now. However, I think that it is not the best approach in the context of docker (containers). We can have a base image (without jdk) and multiple images based on it, each one with a different version of jdk. So, in order to choose one, you just start the corresponding image. The advantage is that users that want a single version do not need to download all of them.

Nevertheless, if you find it useful for development to have a single image with multiple jdk versions, we can add it.

> That's good for me ^^, but just one thing, "does not include any desktop environment", does that mean that we can't run Intellij from the image ? Would it be possible to have IDE like Intellij / eclipse in the docker image ?

You can and cannot. ATM, image `spinalhdl/dev` is to compile and test SpinalHDL in CI flows, where there is no X server available. Therefore, the image does not include neither Xorg nor Intellij itself, so that space and bandwidth requirements are reduced. But there are two alternatives to add them:

- Start a container and install them with `apt`. The drawback is that this needs to be done each time a container is started, because containers are temporal.
- Build a second image, say `spinalhdl/dev:ide`, which includes Intellij on top of `spinal/dev`. I assume that X libraries will be installed as a dependency of the IDE.
 
Just to make it clear. I understand that you want Intellij to develop SpinalHDL. Therefore, it should be added to repo `SpinalHDL/docker`, not here. However, if you want to provide an image with SpinalHDL preinstalled along with Intellij, so that users can develop designs with it, that would be added to the Dockerfile in this PR.

> I created an docker acount : dolu1990
> You can add it :)

It is added now. You should have access to all the settings of the organization, and the repositories.

---

> So, it's fine if the sbt clean test is run only on a single JDK, but at least, an sbt clean compile should run on all JDK >= 8.

Compared to the initial proposal, now the docker execution is an additional job in Travis, but the previous 'native' jobs are not removed. So, ATM, JDK 9, 10 and 11 are used just as before, and JDK 8 is used inside a container.

I think that this approach can make the transition easier. Other JDK versions can be moved to docker images in the future.

Nevertheless, I moved all the `before_install` steps into a function in `travis.sh` and I updated the tools (GHDL, iverilog, cocotb and verilator) so the they match those inside the docker container. Also, YAML anchors are used, instead of Travis' matrix expansion.

@Dolu1990, if you are somehow ok with the implementation, I think that the next step would be to fork [1138-4EB/SpinalHDL-docker](https://github.com/1138-4EB/SpinalHDL-docker) to `SpinalHDL/docker`, and set up docker credentials in Travis, so that images in docker hub are updated from there.

wifasoi commented:
If i can add my 2 cent about intellij-Idea on a container:
I think is more useful to reuse the CI container with a custom entry-point.
This entry-point will launch "sbt run" into the project folder (that you mount before starting the container),
or you can have sbt listen for changes in the mounted folder and trigger a build every time a file changes. So all will be editor agnostic.
Maybe bloop can help for integrating the building process with the IDE, but I don't know it enough to suggest it as solution.

1138-4EB commented:
@wifasoi, that's true, and it is possible with the current implementation, already. However, the *editor agnostic* approach requires some editor to be installed on the host. I understood that @Dolu1990 wanted to have the IDE inside the container, in order to avoid installing it on the host.

Dolu1990 commented:
Hmm or just, having an option to have IDE included or not ?

1138-4EB commented:
> Hmm or just, having an option to have IDE included or not ?

Yes, that's what I commented above. The 'option' is that the user can choose which container/image to use. Image `spinalhdl/dev` will not include any IDE, but `spinalhdl/dev:ide` will. Certainly, because docker images are composed of layers, `spinalhdl/dev:ide` will have all the same layers as `spinalhdl/dev`, plus the IDE.

Furthermore, it is possible to have multiples 'ide' images, say `spinalhdl/dev:intellij`, `spinalhdl/dev:eclipse`, etc. so that each user can pick their favourite, while all of them share the same base resources. We can make profit of docker's multi-stage images to do so. We are already doing it in this PR:

``` dockerfile
FROM spinalhdl/dev as spinalhdl
...
FROM spinalhdl as riscv
```

`spinalhdl/spinalhdl` is based on `spinalhdl/dev`, and `spinalhdl/riscv` is based on `spinalhdl/spinalhdl`.

Nonetheless, as said, I think that tinkering with IDEs and X servers is out of the scope of this PR, which is the very first integration of docker into the project repos. I think we should clean and implement this before we move ahead.

---

Just as a reference of the ideas I'd like to bring later, [eclipse/che](https://www.eclipse.org/che/) is a interesting project for higher-level work environments (such as those that suit SpinalHDL). I gave it a try a year ago (see [/ghdl/ghdl#477](https://github.com/ghdl/ghdl/issues/477#tocome)), and although there are some limitations (see [eclipse/che#3595](https://github.com/eclipse/che/issues/3595#issuecomment-349852819)), I think it can be interesting to tinker with once we have some basic integration done.

wifasoi commented:
I know this is a bit OT:
eclipse che doesn't support scala.. yet... unless I/(or somebody else) add support for metals (the new LSP for scala) and scala/bloop/...
Theia (the web ide part of the eclipse/che ecosystem) can be used instead a full blow eclipse/che eviroment

wifasoi commented:
And yes I agree, the IDE part of this discussion is out of scope of this PR, maybe we can move this discussion in a issue for the future.

Dolu1990 commented:
> Yes, that's what I commented above. The 'option' is that the user can choose which container/image to use. Image `spinalhdl/dev` will not include any IDE, but `spinalhdl/dev:ide` will. Certainly, because docker images are composed of layers, `spinalhdl/dev:ide` will have all the same layers as `spinalhdl/dev`, plus the IDE.

Sound good for me :D

> Nonetheless, as said, I think that tinkering with IDEs and X servers is out of the scope of this PR, which is the very first integration of docker into the project repos. I think we should clean and implement this before we move ahead.

That's right


>Furthermore, it is possible to have multiples 'ide' images, say `spinalhdl/dev:intellij`, `spinalhdl/dev:eclipse`, etc. so that each user can pick their favourite, while all of them share the same base resources. We can make profit of docker's multi-stage images to do so. We are already doing it in this PR:

About the eclipse scala IDE (the base one), i had most of the time bad experiences with it, i'm only working with intellij right now. Is somebody using it the eclipse scala IDE ?



I forked the docker repo here : https://github.com/SpinalHDL/SpinalHDL-docker invited @1138-4EB and @wifasoi in it.


@1138-4EB  @wifasoi can i merge this pull request : D ?

1138-4EB commented:
> @1138-4EB @wifasoi can i merge this pull request : D ?

Not yet, please. There are a couple of minor things to clean. E.g.: `repo: 1138-4EB/SpinalHDL`. I'm reaching to you on Gitter.

1138-4EB commented:
@Dolu1990, I think it is ready to merge now. In the last commit I changed `on.repo` to `SpinalHDL/SpinalHDL` and `on.branch` to `master`. I am sure about the first one, but not so much about the second.

If we leave it like this, docker images will not be updated according to the development branch. So there will be no images available with the development version for users to try. However, all the images will be stable as soon as this PR is merged to master (in the next release).

If we change it to `dev`, docker images will be update after each push to branch `dev`. This means that docker images (spinalhdl/spinalhdl and spinalhdl/riscv) will be 'rolling images' and there will be no stable images for end users. This will be so until we implement a mechanism to double-tag docker images when a release is published (a tagged commit).

Personally, I think that it is better to change it to `dev` and do not advertise docker images to users yet. After testing it for some days/weeks, after implementing the release flow, and after adding the variants with IDEs, we can advertise it all at once.

wifasoi commented on .travis.yml:
Sorry, i didn't summit the review (i'm a big dummy). I'll close this one.

1138-4EB commented on Dockerfile:
```diff
@@ -0,0 +1,28 @@
+FROM spinalhdl/dev as spinalhdl
```
`spinalhdl/spinalhdl` and `spinalhdl/riscv` are ready-to-use images for the end-user. They have a version of SpinalHDL preinstalled, which is known to work because it is tested just before being pushed. The idea is that, for each push to this repo, the following steps are executed:

- Compile and test SpinalHDL in a `spinalhdl/dev`.
- If successful, install SpinalHDL in new image (`spinalhdl/spinalhdl`) and push it to dockerhub.
  - Add RISCV toolchain to the previous one and push it to dockerhub (`spinalhdl/riscv`).

The advantage of including the dockerfile and building the images in this repo, is that images can be tagged when a release is pushed. Otherwise, we would need some additional mechanism to trigger a build in a separate repo.

However, I am not very sure about how are dependencies managed in Scala. Is `sbt clean compile publishLocal` enough to ensure that all the required dependencies are included in the image? I.e., that the container can be used for offline development? Also, is it necessary to keep the git repo or can it be removed after `publishLocal`?

wifasoi replied:
> The advantage of including the dockerfile and building the images in this repo, is that images can be tagged when a release is pushed. Otherwise, we would need some additional mechanism to trigger a build in a separate repo.

My argumentation is based only on personal preference; I like to keep a logical separation between repo(TL;DR: I'm ok to have a dockerfile for spinalhdl, but derivative container should live in another repo). I don't see a easy way to have a separate repo and trigger a build on each release on the other (webhooks can be used, but i feel is more a hack than a solution), so I'm not totally against the inclusion of the dockerfile in the main repo.

> However, I am not very sure about how are dependencies managed in Scala. Is `sbt clean compile publishLocal` enough to ensure that all the required dependencies are included in the image? I.e., that the container can be used for offline development? Also, is it necessary to keep the git repo or can it be removed after `publishLocal`?

It should be enough to launch `sbt clean publishLocal` , and it will populate the ~/.ivy folder (is the cache that sbt use for fetching dependencies) with the necessary files. sbt will fetch the artifacts/files from the ivy cache... so you don't need the source code on the machine.
